topology/probes/ovs_of.go

```diff
@@ -0,0 +1,503 @@
+						probe.Rules[rule.UUID] = rule
+						addRule(rule, probe)
+					}
+				case "DELETED":
+					delete(probe.Rules, event.Rule.UUID)
+					delRule(event.Rule, probe)
+				}
+			} else {
+				logging.GetLogger().Errorf("Error while monitoring %s@%s: %s", probe.Bridge, probe.Host, err.Error())
+			}
+		}
+
+	}()
+	return nil
+}
+
+// NewBridgeProbe creates a probe and launch the active process
+func (o *OvsOfProbe) NewBridgeProbe(host string, bridge string, uuid string) (*BridgeOfProbe, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	address, ok := o.Translation[bridge]
+	if !ok {
+		address = bridge
+	}
+	probe := BridgeOfProbe{
+		Host:       host,
+		Bridge:     bridge,
+		UUID:       uuid,
+		Address:    address,
+		BridgeNode: nil,
+		OvsOfProbe: o,
+		Rules:      make(map[string]*Rule),
+		cancel:     cancel}
+	err := monitor(ctx, &probe)
+	return &probe, err
+}
+
+func (o *OvsOfProbe) makeCommand(commands []string, bridge string, args ...string) ([]string, error) {
+	commandLine := []string{}
+	if o.Sudo {
+		commandLine = []string{"sudo"}
+	}
+	commandLine = append(commandLine, commands...)
+	if strings.HasPrefix(bridge, "ssl:") {
+		if o.sslOk {
+			commandLine = append(commandLine,
+				bridge,
+				"--certificate", o.Certificate,
+				"--ca-cert", o.CA, "--private-key", o.PrivateKey)
+		} else {
+			return commands, errors.New("Certificate, CA and private keys are necessary for communication with switch over SSL")
+		}
+	} else {
+		commandLine = append(commandLine, bridge)
+	}
+	commandLine = append(commandLine, args...)
+	return commandLine, nil
+}
+
+// OnOvsBridgeAdd is called when a bridge is added
+func (o *OvsOfProbe) OnOvsBridgeAdd(uuid string, row *libovsdb.RowUpdate) {
+	o.Lock()
+	defer o.Unlock()
+	bridgeName := row.New.Fields["name"].(string)
+	hostName := o.Host
+	_, ok := o.BridgeProbes[bridgeName]
+	if ok {
+		return
+	}
+	bridgeProbe, err := o.NewBridgeProbe(hostName, bridgeName, uuid)
+	if err == nil {
+		o.BridgeProbes[bridgeName] = bridgeProbe
+	} else {
+		logging.GetLogger().Errorf("Cannot add bridge %s@%s", bridgeName, hostName)
+	}
+}
+
+// OnOvsBridgeDel is called when a bridge is deleted
+func (o *OvsOfProbe) OnOvsBridgeDel(uuid string, row *libovsdb.RowUpdate) {
+	o.Lock()
+	defer o.Unlock()
+	bridgeName := row.New.Fields["name"].(string)
+	bridgeProbe := o.BridgeProbes[bridgeName]
+	bridgeProbe.cancel()
+	delete(o.BridgeProbes, bridgeName)
+}
+
+// NewOvsOfProbe creates a new probe associated to a given graph, root node and host.
+func NewOvsOfProbe(g *graph.Graph, root *graph.Node, host string) *OvsOfProbe {
+	enable := config.GetConfig().GetBool("ovsof.enable")
+	if !enable {
+		return nil
+	}
+	translate := config.GetConfig().GetStringMapString("ovsof.address")
+	cert := config.GetConfig().GetString("ovsof.certificate")
+	pk := config.GetConfig().GetString("ovsof.key")
+	ca := config.GetConfig().GetString("ovsof.ca")
+	sudo := config.GetConfig().GetBool("ovsof.sudo")
```

Comment:
Yes may be not. But I have always wondered if we could find a configuration where root would not be needed as it would be better if we could limit the right of agents to a less powerful user if we want to deploy it on production clouds. At least sudo can be limited to a given command.